File: tests/discord/client.test.ts
import { afterEach, beforeEach, describe, expect, it, mock, spyOn } from "bun:test";
import pino from "pino";
import { sendEmbed } from "../../src/discord/client.js";
import { NonRetryableError, RetryableError } from "../../src/errors/index.js";
import type { DiscordEmbed } from "../../src/types/index.js";

const logger = pino({ level: "silent" });

const testEmbed: DiscordEmbed = {
	title: "Test",
	color: 0x00ff00,
};

describe("sendEmbed", () => {
	let fetchMock: ReturnType<typeof mock>;

	beforeEach(() => {
		fetchMock = mock(() => Promise.resolve(new Response(null, { status: 200 })));
		spyOn(globalThis, "fetch").mockImplementation(fetchMock as unknown as typeof fetch);
	});

	afterEach(() => {
		mock.restore();
	});

	describe("successful requests", () => {
		it("should send embed to webhook successfully", async () => {
			// Arrange
			const webhookUrl = "https://discord.com/api/webhooks/123/token";

			// Act
			await sendEmbed(webhookUrl, testEmbed, logger);

			// Assert
			expect(fetchMock).toHaveBeenCalledTimes(1);
			expect(fetchMock).toHaveBeenCalledWith(
				webhookUrl,
				expect.objectContaining({
					method: "POST",
					headers: { "Content-Type": "application/json" },
				}),
			);
		});

		it("should send correct payload structure", async () => {
			// Arrange
			const webhookUrl = "https://discord.com/api/webhooks/123/token";

			// Act
			await sendEmbed(webhookUrl, testEmbed, logger);

			// Assert
			const call = fetchMock.mock.calls[0] as [string, RequestInit];
			const body = JSON.parse(call[1].body as string);
			expect(body).toEqual({ embeds: [testEmbed] });
		});
	});

	describe("rate limiting (429)", () => {
		it("should throw RetryableError on rate limit", async () => {
			// Arrange
			fetchMock.mockResolvedValue(
				new Response(null, {
					status: 429,
					headers: { "retry-after": "5" },
				}),
			);
			const webhookUrl = "https://discord.com/api/webhooks/123/token";

			// Act & Assert
			await expect(sendEmbed(webhookUrl, testEmbed, logger)).rejects.toBeInstanceOf(RetryableError);
		});

		it("should include retry-after in error context", async () => {
			// Arrange
			fetchMock.mockResolvedValue(
				new Response(null, {
					status: 429,
					headers: { "retry-after": "10" },
				}),
			);
			const webhookUrl = "https://discord.com/api/webhooks/123/token";

			// Act & Assert
			try {
				await sendEmbed(webhookUrl, testEmbed, logger);
			} catch (error) {
				expect(error).toBeInstanceOf(RetryableError);
				expect((error as RetryableError).code).toBe("RATE_LIMITED");
				expect((error as RetryableError).context?.["retryAfter"]).toBe(10000);
			}
		});
	});

	describe("server errors (5xx)", () => {
		it("should retry on server error and throw RetryableError after max retries", async () => {
			// Arrange
			fetchMock.mockResolvedValue(new Response(null, { status: 500 }));
			const webhookUrl = "https://discord.com/api/webhooks/123/token";

			// Act & Assert
			await expect(sendEmbed(webhookUrl, testEmbed, logger)).rejects.toBeInstanceOf(RetryableError);
			expect(fetchMock).toHaveBeenCalledTimes(3);
		});
	});

	describe("client errors (4xx)", () => {
		it("should throw NonRetryableError on client error", async () => {
			// Arrange
			fetchMock.mockResolvedValue(new Response("Bad Request", { status: 400 }));
			const webhookUrl = "https://discord.com/api/webhooks/123/token";

			// Act & Assert
			await expect(sendEmbed(webhookUrl, testEmbed, logger)).rejects.toBeInstanceOf(
				NonRetryableError,
			);
		});

		it("should not retry on 4xx errors", async () => {
			// Arrange
			fetchMock.mockResolvedValue(new Response("Forbidden", { status: 403 }));
			const webhookUrl = "https://discord.com/api/webhooks/123/token";

			// Act & Assert
			try {
				await sendEmbed(webhookUrl, testEmbed, logger);
			} catch {
				// Expected
			}
			expect(fetchMock).toHaveBeenCalledTimes(1);
		});
	});

	describe("network errors", () => {
		it("should throw RetryableError on network failure after retries", async () => {
			// Arrange
			fetchMock.mockRejectedValue(new Error("Network error"));
			const webhookUrl = "https://discord.com/api/webhooks/123/token";

			// Act & Assert
			await expect(sendEmbed(webhookUrl, testEmbed, logger)).rejects.toBeInstanceOf(RetryableError);
			expect(fetchMock).toHaveBeenCalledTimes(3);
		});
	});

	describe("timeout errors", () => {
		it("should throw RetryableError on timeout after max retries", async () => {
			// Arrange
			const abortError = new Error("The operation was aborted");
			abortError.name = "AbortError";
			fetchMock.mockRejectedValue(abortError);
			const webhookUrl = "https://discord.com/api/webhooks/123/token";

			// Act & Assert
			await expect(sendEmbed(webhookUrl, testEmbed, logger)).rejects.toBeInstanceOf(RetryableError);
			expect(fetchMock).toHaveBeenCalledTimes(3);
		});

		it("should retry on timeout before throwing", async () => {
			// Arrange
			const abortError = new Error("The operation was aborted");
			abortError.name = "AbortError";
			fetchMock
				.mockRejectedValueOnce(abortError)
				.mockRejectedValueOnce(abortError)
				.mockResolvedValueOnce(new Response(null, { status: 200 }));
			const webhookUrl = "https://discord.com/api/webhooks/123/token";

			// Act
			await sendEmbed(webhookUrl, testEmbed, logger);

			// Assert
			expect(fetchMock).toHaveBeenCalledTimes(3);
		});
	});

	describe("invalid URLs", () => {
		it("should handle invalid webhook URL gracefully", async () => {
			// Arrange
			fetchMock.mockResolvedValue(new Response(null, { status: 200 }));
			const invalidUrl = "not-a-valid-url";

			// Act
			await sendEmbed(invalidUrl, testEmbed, logger);

			// Assert
			expect(fetchMock).toHaveBeenCalledTimes(1);
		});
	});
});

File: test/notification_service/discord/client_test.exs
defmodule NotificationService.Discord.ClientTest do
  use ExUnit.Case, async: true

  alias NotificationService.Discord.Client
  alias NotificationService.{RetryableError, NonRetryableError}

  @embed %{title: "Test", color: 0x00FF00}

  describe "send_embed/2" do
    test "returns :ok on successful 200 response" do
      # Arrange
      bypass = Bypass.open()

      Bypass.expect_once(bypass, "POST", "/webhooks/test/token", fn conn ->
        Plug.Conn.send_resp(conn, 200, "")
      end)

      # Act
      result = Client.send_embed(endpoint_url(bypass, "/webhooks/test/token"), @embed)

      # Assert
      assert result == :ok
    end

    test "returns :ok on 204 no content response" do
      # Arrange
      bypass = Bypass.open()

      Bypass.expect_once(bypass, "POST", "/webhooks/test/token", fn conn ->
        Plug.Conn.send_resp(conn, 204, "")
      end)

      # Act
      result = Client.send_embed(endpoint_url(bypass, "/webhooks/test/token"), @embed)

      # Assert
      assert result == :ok
    end

    test "returns NonRetryableError on 400 client error" do
      # Arrange
      bypass = Bypass.open()

      Bypass.expect_once(bypass, "POST", "/webhooks/test/token", fn conn ->
        Plug.Conn.send_resp(conn, 400, Jason.encode!(%{message: "Bad Request"}))
      end)

      # Act
      result = Client.send_embed(endpoint_url(bypass, "/webhooks/test/token"), @embed)

      # Assert
      assert {:error, %NonRetryableError{code: "DISCORD_CLIENT_ERROR"}} = result
    end

    test "returns NonRetryableError on 404 not found" do
      # Arrange
      bypass = Bypass.open()

      Bypass.expect_once(bypass, "POST", "/webhooks/test/token", fn conn ->
        Plug.Conn.send_resp(conn, 404, "Not Found")
      end)

      # Act
      result = Client.send_embed(endpoint_url(bypass, "/webhooks/test/token"), @embed)

      # Assert
      assert {:error, %NonRetryableError{}} = result
    end

    test "retries on 500 server error and returns RetryableError after exhaustion" do
      # Arrange
      bypass = Bypass.open()

      # Expect 3 attempts (initial + 2 retries)
      Bypass.expect(bypass, "POST", "/webhooks/test/token", fn conn ->
        Plug.Conn.send_resp(conn, 500, "Internal Server Error")
      end)

      # Act
      result = Client.send_embed(endpoint_url(bypass, "/webhooks/test/token"), @embed)

      # Assert
      assert {:error, %RetryableError{code: "DISCORD_SERVER_ERROR"}} = result
    end

    test "retries on 429 rate limit with retry-after header" do
      # Arrange
      bypass = Bypass.open()
      agent = start_supervised!({Agent, fn -> 0 end})

      Bypass.expect(bypass, "POST", "/webhooks/test/token", fn conn ->
        count = Agent.get_and_update(agent, fn n -> {n, n + 1} end)

        if count == 0 do
          conn
          |> Plug.Conn.put_resp_header("retry-after", "0.001")
          |> Plug.Conn.send_resp(429, Jason.encode!(%{message: "Rate limited"}))
        else
          Plug.Conn.send_resp(conn, 204, "")
        end
      end)

      # Act
      result = Client.send_embed(endpoint_url(bypass, "/webhooks/test/token"), @embed)

      # Assert
      assert result == :ok
      assert Agent.get(agent, & &1) == 2
    end

    test "returns RetryableError on connection refused" do
      # Arrange - use a port that's not listening
      bypass = Bypass.open()
      port = bypass.port
      Bypass.down(bypass)

      # Act
      result = Client.send_embed("http://localhost:#{port}/webhooks/test/token", @embed)

      # Assert
      assert {:error, %RetryableError{code: "DISCORD_TRANSPORT_ERROR"}} = result
    end
  end

  defp endpoint_url(bypass, path) do
    "http://localhost:#{bypass.port}#{path}"
  end
end
